crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3972,6 +3973,47 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn nonlocal_name_reference_skips_annotation_only_assignment() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+            def f():
+                x = 1
+                def g():
+                    // it's pretty weird to have an annotated assignment in a function where the
+                    // name is otherwise not defined; maybe should be an error?
+                    x: int
```

Comment:
I think this should be an error because the annotation is incorrect. The annotation should be `x: int | None` because it isn't guaranteed to be initialized in all paths.

Reply:
`int | None` would be much less accurate here. `int | None` means "the value could be an instance of `int`, or it could be the `None` singleton". But in no path is `x` assigned to the `None` singleton here, so it's impossible for the value of `x` to ever be `None`.

There's no way in Python to express with type annotations the idea that "this variable may be undefined". We use `Type::Unbound` internally to track whether a variable has been defined in all branches or not, but that's not a specified type in Python's typing system that users are able to utilise in their annotations

Reply:
Hmm, that's a rather interesting choice in python's type system.

Reply:
There have been proposals to improve this, but they've met with a somewhat cool reception. Some people have argued that allowing users to annotate variables as being only possibly defined would encourage users to write more unsafe code. https://mail.python.org/archives/list/typing-sig@python.org/thread/K3ME5IZIV3Q4PTH4PK7YHYJFGBZOLOJK/#K3ME5IZIV3Q4PTH4PK7YHYJFGBZOLOJK

Reply:
I think I tend to agree with Eric on this one; explicit annotations for undefined would just lead to a bunch more ugly code using `hasattr`, since that's how you'd have to narrow away the maybe-undefinedness. It's better to just have a diagnostic on use of a maybe-undefined name, and on failure to consistently initialize instance attributes.